### ISSUE TITLE
Added DBGP command `source` support

### ISSUE THREAD

**Issue by thqby** [zero-plusplus/vscode-autohotkey-debug#208]
> The protocol reserves the URI scheme 'dbgp' for all virtual files generated and maintained by language engines. Such virtual files are usually managed by a language engine for dynamic code blocks, i.e. code created at runtime, without an association with a regular file. Any IDE seeing an URI with the 'dbgp' scheme has to use the 'source' command (See section 7.14) to obtain the contents of the file from the engine responsible for that URI.
All URIs in that scheme have the form: `dbgp:engine-specific-identifier`
http://xdebug.org/docs/dbgp#dynamic-code-and-virtual-files

[`source` command](http://xdebug.org/docs/dbgp#source)
>The body of the request is the URI (retrieved from the stack info), the body of the response is the data contents of the URI. If the file uri is not provided, then the file for the current context is returned.

|  option  |   |
|  ----  | ----  |
|-b | begin line (optional)|
|-e | end line (optional)|
|-f | file URI |

IDE command: `source -i transaction_id -f fileURI`

**Comment by zero-plusplus** [comment]
Implement the source command to do what?

I did not implement this command intentionally because I understood it to be for getting the source code text from a URI. (Because it can be implemented using vscode).

**Comment by thqby** [comment]
ahk can include resource files of ahk.exe, such as `#include *libname`. These resources are not a real file.

In addition, ahk_h has the ability to generate code dynamically, which can be obtained through the `source` command.

**Comment by zero-plusplus** [comment]
What specific features can it provide to users?

**Comment by thqby** [comment]
For ahk_H, it will be able to debug `ahkexec`, `addscript`, `newthread` without generating temporary files and lib files in resources.

`dbgp:engine-specific-identifier` is not supported in ahk_L.

**Comment by zero-plusplus** [comment]
I only know that ahk_H can dynamically evaluate AutoHotkey scripts, does that mean that I will be able to debug those dynamically evaluated script?

**Comment by thqby** [comment]
Yes.

**Comment by zero-plusplus** [comment]
I've heard good things about that. I had given up on implementing it because I thought it would not be possible.

But I thought the `source` command was just to get the source code, how can I use this to implement that?

**Comment by thqby** [comment]
Vscode can preview virtual files, like git's file history. Can be generated with the source code obtained, but I don't know how to associate with the debugger.

`TextDocumentContentProvider` can be used to generate a read-only TextDocument.
`export function registerTextDocumentContentProvider(scheme: string, provider: TextDocumentContentProvider): Disposable;`

**Comment by zero-plusplus** [comment]
https://github.com/HotKeyIt/ahkdll/blob/master/source/Debugger.cpp#L1948

I am not familiar with c++  so I could be wrong, but it still seems that the `source` command only returns the loaded source code.

So I don't think it is possible to debug dynamically evaluated script with it.

**Comment by thqby** [comment]
https://github.com/thqby/AutoHotkey_H/blob/alpha/source/Debugger.cpp#L2594

https://github.com/thqby/AutoHotkey_H/blob/alpha/source/Debugger.cpp#L1967

This is a new feature I added.

**Comment by zero-plusplus** [comment]
For now I only support AutoHotkey_L from Lexikos and AutoHotkey_H (ahkdll) from HotkeyIt, the most famous fork.

This is because incorporating the features of individual forks would make the source code too complex.

Therefore, I cannot accept this request.

However, since debugging dynamically evaluated script is a great feature, can you negotiate to allow us to merge your changes into HotkeyIt's AutoHotkey_H?

I would gladly work on implementing this feature when it is incorporated into AutoHotkey_H.

**Comment by thqby** [comment]
Certainly. 
This is a part of the dbgp command. The implementation is also very simple and will not complicate the source code.

**Comment by zero-plusplus** [comment]
Doesn't your example not debug dynamically evaluated scripts but just display that script in the editor?

The important thing is to make sure that the debugger itself has the features to run, stop, or step through dynamically evaluated script.

**Comment by thqby** [comment]
It should be that just provide the content to vscode.
The debugger runs line by line, but nothing is displayed in the editor.

**Comment by zero-plusplus** [comment]
I apologize in advance because this is an example that will make you feel bad, but let's assume, for example, that your AutoHotkey_H has a security vulnerability or is itself malware.

If a user of my extension is victimized by it, it may not forgive me for supporting it and introducing it. Because by doing so, I would be responsible for it.

So it is very risky for me to implement a feature that is only included in your AutoHotkey_H.

Including me, vscode extensions are secured by microsoft checks, and AutoHotkey_L and H are secured by the community.

It means that it is very difficult for software that is not to do so to gain trust. This includes me. That's why I used the metaphor that even if you were Bill Gates, I would say no.

---

I ran your example without any modifications to the this extension (v1.11.0) and was able to debug the dynamically evaluated script.
